src/cache.rs

```diff
@@ -100,124 +138,167 @@ pub fn id() -> usize {
 /// Comemo's cache is thread-local, meaning that this only evicts this thread's
 /// cache.
 pub fn evict(max_age: usize) {
-    CACHE.with(|cache| {
-        let mut cache = cache.borrow_mut();
-        cache.map.retain(|_, entries| {
-            entries.retain_mut(|entry| {
-                entry.age += 1;
-                entry.age <= max_age
-            });
-            !entries.is_empty()
-        });
-    });
-    ACCELERATOR.with(|accelerator| accelerator.borrow_mut().clear());
+    CACHES.read().iter().for_each(|fun| fun(max_age));
+
+    // Evict all accelerators.
+    let mut accelerators = ACCELERATORS.write();
+
+    // Force the ID to be loaded.
+    let id = ID.load(Ordering::Acquire);
+
+    // Update the offset.
+    accelerators.0 = id;
+
+    // Clear all accelerators while keeping the memory allocated.
+    accelerators.1.iter_mut().for_each(|accelerator| {
+        accelerator.lock().clear();
+    })
 }
 
 /// The global cache.
-#[derive(Default)]
-struct Cache {
-    /// Maps from function IDs + hashes to memoized results.
-    map: HashMap<(TypeId, u128), Vec<CacheEntry>>,
-    /// Whether the last call was a hit.
-    last_was_hit: bool,
+pub struct Cache<C, Out> {
+    /// Maps from hashes to memoized results.
+    entries: HashMap<u128, Vec<CacheEntry<C, Out>>>,
+}
+
+impl<C, Out> Default for Cache<C, Out> {
+    fn default() -> Self {
+        Self { entries: HashMap::new() }
+    }
 }
 
-impl Cache {
+impl<C, Out: 'static> Cache<C, Out> {
+    /// Create an empty cache.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Evict all entries whose age is larger than or equal to `max_age`.
+    pub fn evict(&mut self, max_age: usize) {
+        self.entries.retain(|_, entries| {
+            entries.retain(|entry| {
+                let age = entry.age.fetch_add(1, Ordering::Acquire);
```

Comment:
Since we have `&mut self` access to the Cache, I believe no atomic operations are necessary here. Just [`.get_mut()`](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicUsize.html#method.get_mut) on the atomic should work.